File: packages/kotlin/ttr-semantics/src/main/kotlin/org/tatrman/ttr/semantics/semanticsblock/Vocabulary.kt
// SPDX-License-Identifier: Apache-2.0
package org.tatrman.ttr.semantics.semanticsblock

/**
 * Grounding Phase 1 (grammar 4.2) — the closed `semantics { … }` vocabulary.
 *
 * NORMATIVE table, mirroring `packages/semantics/src/semantics-block/vocabulary.ts`
 * and `docs/features/semantics-block/README.md` §Vocabulary exactly. It is the
 * cross-repo sync key with ai-platform's closed proto enums: the vocabulary here
 * and the proto enums version TOGETHER via [SEMANTICS_VOCABULARY_VERSION].
 */
object Vocabulary {
    /**
     * Cross-repo sync key — bumps in lock-step with ai-platform's proto enums. v2 (MD dot-path S5C-B.4)
     * adds the journal-role family (`valid_flag`, `version`, `authored_by`, `written_at`) — the closed
     * six-role set of contracts §12 R30; `valid_from`/`valid_to` were already present and are reused. The
     * shared `AttributeSemanticRole` proto-enum promotion (ids 60–63) is the cross-repo half, coordinated
     * at the metadata-proto release (ai-platform §4) — out of this repo.
     */
    const val SEMANTICS_VOCABULARY_VERSION: Int = 2

    /** Entity/table kinds (`kind:`). */
    val ENTITY_KINDS: List<String> = listOf("period_table", "calendar", "poi", "fx_rate")

    /** One extra key (besides `role`) a role accepts, with its reference kind. */
    data class ExtraKey(
        val key: String,
        val kind: RefKind,
        val required: Boolean,
    )

    enum class RefKind { EntityRef, AttrRef, StringLit }

    /** The type-family a role's attribute/column must declare. */
    enum class TypeConstraint { Date, Text, Numeric }

    data class RoleSpec(
        val extraKeys: List<ExtraKey> = emptyList(),
        val typeConstraint: TypeConstraint? = null,
    )

    /** Attribute/column roles (`role:`) → their extra keys + type constraint. */
    val ATTRIBUTE_ROLES: Map<String, RoleSpec> =
        linkedMapOf(
            "period_start" to RoleSpec(typeConstraint = TypeConstraint.Date),
            "period_end" to RoleSpec(typeConstraint = TypeConstraint.Date),
            "period_code" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("code_format", RefKind.StringLit, required = false)),
                    typeConstraint = TypeConstraint.Text,
                ),
            "event_date" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("period", RefKind.EntityRef, required = false)),
                    typeConstraint = TypeConstraint.Date,
                ),
            "document_date" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("period", RefKind.EntityRef, required = false)),
                    typeConstraint = TypeConstraint.Date,
                ),
            "posting_date" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("period", RefKind.EntityRef, required = false)),
                    typeConstraint = TypeConstraint.Date,
                ),
            "due_date" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("period", RefKind.EntityRef, required = false)),
                    typeConstraint = TypeConstraint.Date,
                ),
            "valid_from" to RoleSpec(typeConstraint = TypeConstraint.Date),
            "valid_to" to RoleSpec(typeConstraint = TypeConstraint.Date),
            // Journal-role family (S5C-B.4, contracts §12 R30 · MDS8) — technical columns of a journaled
            // cubelet's backing table. `valid_flag` is boolean (no numeric/text/date family — left
            // unconstrained); `version` int, `authored_by` varchar, `written_at` timestamp (date family).
            "valid_flag" to RoleSpec(),
            "version" to RoleSpec(typeConstraint = TypeConstraint.Numeric),
            "authored_by" to RoleSpec(typeConstraint = TypeConstraint.Text),
            "written_at" to RoleSpec(typeConstraint = TypeConstraint.Date),
            "calendar_date" to RoleSpec(typeConstraint = TypeConstraint.Date),
            "geo_lat" to RoleSpec(typeConstraint = TypeConstraint.Numeric),
            "geo_lon" to RoleSpec(typeConstraint = TypeConstraint.Numeric),
            "geo_point" to RoleSpec(typeConstraint = TypeConstraint.Text),
            "amount" to
                RoleSpec(
                    extraKeys = listOf(ExtraKey("currency", RefKind.AttrRef, required = false)),
                    typeConstraint = TypeConstraint.Numeric,
                ),
            "amount_domestic" to RoleSpec(typeConstraint = TypeConstraint.Numeric),
            "currency_code" to RoleSpec(typeConstraint = TypeConstraint.Text),
            "fx_from_currency" to RoleSpec(typeConstraint = TypeConstraint.Text),
            "fx_to_currency" to RoleSpec(typeConstraint = TypeConstraint.Text),
            "fx_rate" to RoleSpec(typeConstraint = TypeConstraint.Numeric),
        )

    /** A single required-role clause in a kind's completeness rule (exactly `count` times). */
    data class CompletenessClause(
        val role: String,
        val count: Int = 1,
    )

    /**
     * Per-kind completeness rules (validated on the owning entity/table). `poi` is
     * special (geo_point XOR lat/lon pair) — handled in the validator, so it maps
     * to an empty list here.
     */
    val KIND_COMPLETENESS: Map<String, List<CompletenessClause>> =
        mapOf(
            "period_table" to
                listOf(
                    CompletenessClause("period_start"),
                    CompletenessClause("period_end"),
                    CompletenessClause("period_code"),
                ),
            "calendar" to listOf(CompletenessClause("calendar_date")),
            "poi" to emptyList(),
            "fx_rate" to
                listOf(
                    CompletenessClause("fx_from_currency"),
                    CompletenessClause("fx_to_currency"),
                    CompletenessClause("fx_rate"),
                ),
        )

    val ALL_ROLES: List<String> = ATTRIBUTE_ROLES.keys.toList()

    /** The keys legal on an entity/table `semantics` block. */
    val ALL_ENTITY_KEYS: List<String> = listOf("kind")
}

File: packages/semantics/src/semantics-block/vocabulary.ts
// SPDX-License-Identifier: Apache-2.0
// Grounding Phase 1 (grammar 4.2) — the closed `semantics { … }` vocabulary.
//
// This table is NORMATIVE and mirrors `docs/features/semantics-block/README.md`
// §Vocabulary exactly. It is the cross-repo sync key with ai-platform's closed
// proto enums (`cz.dfpartner.metadata.v1` AttributeSemantics / EntitySemantics,
// feature-grounding-contracts.md §4): the vocabulary here and the proto enums
// version TOGETHER via SEMANTICS_VOCABULARY_VERSION (the md-catalog
// MD_CATALOG_VERSION precedent). Bump it whenever a role/kind is added or a
// signature changes, and cut the matching proto release in lock-step.

/** Cross-repo sync key — bumps in lock-step with ai-platform's proto enums. */
export const SEMANTICS_VOCABULARY_VERSION = 1 as const;

/** The type-family a role's attribute/column must declare. */
export type TypeConstraint = 'date' | 'text' | 'numeric';

/** Entity/table kinds (`kind:`). */
export const ENTITY_KINDS = ['period_table', 'calendar', 'poi', 'fx_rate'] as const;
export type EntityKind = (typeof ENTITY_KINDS)[number];

/** Cross-reference keys a role may carry beyond `role:` itself. */
export interface RoleSpec {
  /** Optional extra keys (besides `role`) this role accepts, with their kind. */
  readonly extraKeys: ReadonlyArray<{ key: string; kind: 'entityRef' | 'attrRef' | 'string'; required: boolean }>;
  /** The declared-type family the attribute/column must have (undefined = any). */
  readonly typeConstraint?: TypeConstraint;
}

/**
 * Attribute/column roles (`role:`) → their extra keys + type constraint. Mirrors
 * README §Vocabulary's role table 1:1.
 */
export const ATTRIBUTE_ROLES: Readonly<Record<string, RoleSpec>> = {
  period_start: { extraKeys: [], typeConstraint: 'date' },
  period_end: { extraKeys: [], typeConstraint: 'date' },
  period_code: { extraKeys: [{ key: 'code_format', kind: 'string', required: false }], typeConstraint: 'text' },
  event_date: { extraKeys: [{ key: 'period', kind: 'entityRef', required: false }], typeConstraint: 'date' },
  document_date: { extraKeys: [{ key: 'period', kind: 'entityRef', required: false }], typeConstraint: 'date' },
  posting_date: { extraKeys: [{ key: 'period', kind: 'entityRef', required: false }], typeConstraint: 'date' },
  due_date: { extraKeys: [{ key: 'period', kind: 'entityRef', required: false }], typeConstraint: 'date' },
  valid_from: { extraKeys: [], typeConstraint: 'date' },
  valid_to: { extraKeys: [], typeConstraint: 'date' },
  calendar_date: { extraKeys: [], typeConstraint: 'date' },
  geo_lat: { extraKeys: [], typeConstraint: 'numeric' },
  geo_lon: { extraKeys: [], typeConstraint: 'numeric' },
  geo_point: { extraKeys: [], typeConstraint: 'text' },
  amount: { extraKeys: [{ key: 'currency', kind: 'attrRef', required: false }], typeConstraint: 'numeric' },
  amount_domestic: { extraKeys: [], typeConstraint: 'numeric' },
  currency_code: { extraKeys: [], typeConstraint: 'text' },
  fx_from_currency: { extraKeys: [], typeConstraint: 'text' },
  fx_to_currency: { extraKeys: [], typeConstraint: 'text' },
  fx_rate: { extraKeys: [], typeConstraint: 'numeric' },
} as const;

export type AttributeRole = keyof typeof ATTRIBUTE_ROLES;

/**
 * A single required-role clause in a kind's completeness rule: the role must
 * appear exactly `count` times among the owner's attributes/columns.
 */
export interface CompletenessClause {
  readonly role: AttributeRole;
  readonly count: 1;
}

/**
 * Per-kind completeness rules (validated on the owning entity/table). `poi` is
 * special (geo_point XOR lat/lon pair) — handled in the validator, not as a flat
 * clause list — so it maps to an empty list here and is documented as such.
 */
export const KIND_COMPLETENESS: Readonly<Record<EntityKind, ReadonlyArray<CompletenessClause>>> = {
  period_table: [
    { role: 'period_start', count: 1 },
    { role: 'period_end', count: 1 },
    { role: 'period_code', count: 1 },
  ],
  calendar: [{ role: 'calendar_date', count: 1 }],
  // poi: exactly one geo_point XOR (one geo_lat AND one geo_lon) — the XOR is not
  // expressible as a flat count list, so the validator special-cases it (210).
  poi: [],
  fx_rate: [
    { role: 'fx_from_currency', count: 1 },
    { role: 'fx_to_currency', count: 1 },
    { role: 'fx_rate', count: 1 },
    // valid_from/valid_to are optional as a pair (211), not required — so absent here.
  ],
} as const;

export const ALL_ROLES: ReadonlyArray<string> = Object.keys(ATTRIBUTE_ROLES);

/** The keys legal at all: `role` plus every role's extra keys, deduped. */
export const ALL_ATTRIBUTE_KEYS: ReadonlyArray<string> = [
  'role',
  ...new Set(Object.values(ATTRIBUTE_ROLES).flatMap((r) => r.extraKeys.map((k) => k.key))),
];

/** The keys legal on an entity/table `semantics` block. */
export const ALL_ENTITY_KEYS: ReadonlyArray<string> = ['kind'];
